addons/sale_margin/models/sale_order.py

```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+import odoo.addons.decimal_precision as dp
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    margin = fields.Float(compute='_product_margin', digits=dp.get_precision('Product Price'), store=True)
+    purchase_price = fields.Float(string='Cost Price', digits=dp.get_precision('Product Price'))
+
+    @api.onchange('product_id', 'product_uom_qty')
+    def product_id_change_margin(self):
+        if self.order_id.pricelist_id:
```

Comment:
Please avoid refactoring when porting. This one was already in new API so shouldn't have required any changes.